clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -424,9 +424,63 @@ caseCase _ e = return e
 {-# SCC caseCase #-}
 
 -- | Inline function with a non-representable result if it's the subject
--- of a Case-decomposition
+-- of a Case-decomposition. It's a custom topdown traversal that -for efficiency
+-- reasons- does not explore alternative of cases whose subject triggered an
+-- 'inlineNonRepWorker'.
 inlineNonRep :: HasCallStack => NormRewrite
-inlineNonRep _ e@(Case scrut altsTy alts)
+inlineNonRep _ctx = fmap snd . go
+ where
+  go :: HasCallStack => Term -> NormalizeSession (Bool, Term)
```

Comment:
I don't like that you return a `Bool` here which is just thrown away. `go` should just return the final `Term` IMO